### ISSUE TITLE
Better support for credential rotation

### ISSUE THREAD

**Issue by sgallag-insta** [doorkeeper-gem/doorkeeper#1675]
Right now credential rotation while using Doorkeeper seems a little cumbersome. From my POV, there are currently 2 major approaches that can be taken:

1. Only the secret is rotated. This approach seems more elegant, but the major drawback here is that client applications cannot authenticate until they have updated their credentials in the code.
2. The entire OAuth application is rotated. This allows for a grace period during which time new credentials can be added/tested while the old credentials remain active. However, this method is not without consequences. Assuming we do not purge all tokens during the rotation, if the client wants to revoke a token during the transition period, they may end up attempting to revoke a token created by a different OAuth application (not allowed). Also if anyone is using the oauth_application_id or uid for metrics they have to take this transition into account.

My suggestion/feature request here would be to add additional support for approach 1. Perhaps we could have an "old_secret" field that remains active for a period of time after the secret field is updated. Thoughts?

**Comment by vix633** [comment]
It's a really large missing piece here

**Comment by ClaytonPassmore** [comment]
I agree that having an additional field to store the "old" secret is a good approach.

When it comes to rotating out the old secret, you could definitely use some kind of time-based expiration like you mentioned (probably with the help of an additional field to track the expiration). However, I think there's a more simple approach that would offer a bit more flexibility: You could let the old secret continue to function until it is removed from the application by the application owner.

I'm imagining a flow like this:

1. App owner generates a new secret, the old secret is retained in the `old_secret` field
2. App owner updates their environment to use new secret
3. App owner removes the old secret from the application

This gives the application owner as much time as they need to update their environment. The nice thing about this approach is that if _you_ want to add in a custom time-based expiration mechanism, you are free to do so with some kind of background job. Or you could implement a passive expiration approach by customizing the application class and adding your own additional column and accessor to "ignore" the old secret once it has expired, similar to the solution that was initially proposed.
